packages/paymasters/src/PermitPaymasterUtils.ts

```diff
@@ -0,0 +1,219 @@
+  readonly types: Types
+  readonly domain: EIP712Domain
+  readonly primaryType: string
+  readonly message: any
+
+  constructor (
+    chainId: number,
+    permitType: PermitType,
+    domain: EIP712Domain,
+    permit: PermitInterfaceDAI | PermitInterfaceEIP2612,
+    eip712DomainType: EIP712TypeProperty[] = EIP712DomainType) {
+    this.types = {
+      EIP712Domain: eip712DomainType,
+      Permit: permitType
+    }
+    this.domain = domain
+    this.primaryType = 'Permit'
+    // in the signature, all "request" fields are flattened out at the top structure.
+    // other params are inside "relayData" sub-type
+    this.message = {
+      ...permit
+    }
+  }
+}
+
+export async function signAndEncodeDaiPermit (
+  holder: Address,
+  spender: Address,
+  token: Address,
+  expiry: IntString,
+  web3Input: Web3,
+  domainSeparator: EIP712Domain = getDaiDomainSeparator(),
+  forceNonce?: number,
+  skipValidation = false
+): Promise<PrefixedHexString> {
+  const web3 = new Web3(web3Input.currentProvider)
+  const DaiContract = TruffleContract({
+    contractName: 'DAIPermitInterface',
+    abi: daiPermitAbi.abi
+  })
+
+  DaiContract.setProvider(web3.currentProvider, undefined)
+  const daiInstance = await DaiContract.at(token)
+  const nonce = forceNonce ?? await daiInstance.nonces(holder)
+  const chainId = await web3.eth.getChainId()
+  const permit: PermitInterfaceDAI = {
+    // TODO: not include holder as 'from', not pass 'from'
+    from: holder,
+    holder,
+    spender,
+    nonce,
+    expiry,
+    allowed: true
+  }
+  const dataToSign = new TypedPermit(
+    chainId,
+    PermitTypeDai,
+    domainSeparator,
+    permit
+  )
+  const signature = await getEip712Signature(
+    web3,
+    dataToSign
+  )
+  const { r, s, v } = hex2signatureRSV(signature)
+  // we use 'estimateGas' to check against the permit method revert (hard to debug otherwise)
+  if (!skipValidation) {
+    await daiInstance.contract.methods.permit(holder, spender, nonce, expiry, true, v, r, s).estimateGas()
+  }
+  return daiInstance.contract.methods.permit(holder, spender, nonce, expiry, true, v, r, s).encodeABI()
+}
+
+export async function signAndEncodeEIP2612Permit (
+  owner: Address,
+  spender: Address,
+  token: Address,
+  value: string,
+  deadline: string,
+  web3Input: Web3,
+  domainSeparator: EIP712Domain,
+  domainType?: EIP712TypeProperty[],
+  forceNonce?: number,
+  skipValidation = false
+): Promise<PrefixedHexString> {
+  const web3 = new Web3(web3Input.currentProvider)
+  const EIP2612Contract = TruffleContract({
+    contractName: 'EIP2612Contract',
+    abi: eip2612PermitAbi.abi
+  })
+
+  EIP2612Contract.setProvider(web3.currentProvider, undefined)
+  const eip2612TokenInstance = await EIP2612Contract.at(token)
+  const nonce = forceNonce ?? await eip2612TokenInstance.nonces(owner)
+  const chainId = await web3.eth.getChainId()
+  const permit: PermitInterfaceEIP2612 = {
+    // TODO: not include holder as 'from', not pass 'from'
```

Comment:
'from' field is not described in the EIP-712 type declaration of the Permit. It is not being signed. The reason to include it is that our 'sign712...' method does not accept the 'from' address as the parameter but reads it from the message itself. 
It does not cause any problem today because as we have seen Metamask does not complain and presents 'from' as part of the message being signed. I don't want to change dozens of calls to 'sign712' today, but will do in the near future if we decide to move forward with this permit paymaster somehow.